internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,131 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+// CCGState stores the id used when creating a ccg instance. Used when
+// closing a container to be able to release the instance.
+type CCGState struct {
+	// ID of container that instance belongs to.
+	id string
+}
+
+// Release calls into hcs to remove the ccg instance. These do not get cleaned up automatically
+// they MUST be explicitly removed with a call to ModifyServiceSettings. The instances will persist
+// unless vmcompute.exe exits or they are removed manually as done here.
+func (instance *CCGState) Release(ctx context.Context) error {
+	if err := removeCredentialGuard(ctx, instance.id); err != nil {
+		log.G(ctx).WithError(err).WithField("containerID", instance.id).Warn("failed to remove Container Credential Guard instance")
+		return err
+	}
+	return nil
+}
+
+// CreateCredentialGuard creates a container credential guard and returns the state object to be placed in a v2 container doc.
+func CreateCredentialGuard(ctx context.Context, ID, credSpec string, uvm bool) (*hcsschema.ContainerCredentialGuardState, error) {
+	log.G(ctx).WithField("containerID", ID).Debug("creating container credential guard")
+	// V2 schema ccg setup a little different as its expected to be passed
+	// through all the way to the gcs. Can no longer be enabled just through
+	// a single property. The flow is as follows
+	// ------------------------------------------------------------------------
+	// 1. Call HcsModifyServiceSettings with a ModificationRequest set with a
+	// ContainerCredentialGuardAddInstanceRequest. This is where the cred spec
+	// gets passed in. Transport either "LRPC" (Argon) or "HvSocket" (Xenon).
+	// 2. Query the instance with a call to HcsGetServiceProperties with the
+	// PropertyType "ContainerCredentialGuard". This will return all instances
+	// 3. Parse for the id of our container to find which one correlates to the
+	// container we're building the doc for, then add to the v2 doc.
+	// 4. If xenon container the hvsocketconfig will need to be attached BEFORE
+	// being created. It must be in the doc itself as we do not support hot adding
+	// service table entries. This is currently a blocker for adding support for
+	// hyper-v gmsa.
+	transport := "LRPC"
+	if uvm {
+		transport = "HvSocket"
```

Comment:
We should just return a not supported error here.

Reply:
I'm returning a not supported error when creating the container doc, would you rather it be here? https://github.com/microsoft/hcsshim/blob/d7cf69709fd4817b1f24e8d47631fc6f9f60b1d0/internal/hcsoci/hcsdoc_wcow.go#L171

Reply:
Sorry I meant to respond on this again, but see you already changed it. I suppose it's okay to not throw the error here since technically we set up the CCG instance correctly for hypervisor isolated, it's just the UVM config that isn't correct. However since you already changed it I think it's fine this way too.

Reply:
No worries :)